v2/internal/pkg/operator/common.go

```diff
@@ -161,8 +162,22 @@ func (o OperatorCollector) prepareD2MCopyBatch(images map[string][]v2alpha1.Rela
 			}
 
 			// add the tag for src and dest
-			switch {
 			// applies only to catalogs
+			switch {
+			// OCPBUGS-52470
+			case img.Type == v2alpha1.TypeOperatorCatalog && len(img.TargetTag) == 0 && imgSpec.IsImageByDigestOnly():
+				// find the tag for for the catalog that was mirrored via digest
+				imageIndexDir := filepath.Join(o.Opts.Global.WorkingDir, operatorCatalogsDir, imgSpec.ComponentName(), imgSpec.Digest)
+				filteredCatalogsDir := filepath.Join(imageIndexDir, operatorCatalogFilteredDir)
+				entries, err := os.ReadDir(filteredCatalogsDir)
+				if err != nil {
+					return result, fmt.Errorf("could not read filtered catalog directory for image %q: %w", img.Image, err)
+				}
+				if len(entries) > 1 {
+					return result, fmt.Errorf("digest-based catalog directory %q structure is incorrect: expected 1 entry got %d", filteredCatalogsDir, len(entries))
+				}
+				src = src + ":" + entries[0].Name()
```

Comment:
```suggestion
				// find the tag for for the catalog that was mirrored via digest
				var op v2alpha1.Operator
				for _, ope := range o.Config.Mirror.Operators {
					if ope.Catalog == img.Image {
						op = ope
					}
				}
				if op.Catalog == "" {
					return result, fmt.Errorf("could not identify filtered catalog tag for image %q: no catalog matched %q in the imageSetConfiguration", img.Image, img.Image)
				}
				dof, err := digestOfFilter(op)
				if err != nil {
					return result, fmt.Errorf("could not identify filtered catalog tag for image %q: %w", img.Image, err)
				}
				src = src + ":" + dof
```

Hi @lmzuccarelli,
what I was trying to say in the previous comment is that len(entries) will rarely remain equal to 1, as users might do several mirrorings for the same catalog, eventually with different criteria. 
This snippet shows a bit what I meant, even though it' s not the perfect solution.
We could have this value prepared and set in the relatedImage structure for direct use, before the code reaches `prepareD2MCopyBatch`

Reply:
@sherine-k - hhhmm I disgaree -  we have this type of structure  i.e here is an example 

```
<mirror-dir>/working-dir/operator-catalogs/redhat-operator-index/e3368fb4222d7586a1505748f383fa711b2a2baa231dd02c02b974fd19ade942/filtered-catalogs/2124f1daa35027087f1c61e6631ae521
```
How would it be possible to have multiple entries when we are already in this complex for loop  i.e in lines 

```
	for _, relatedImgs := range images {
		for _, img := range relatedImgs {

```
There should only be one  sha digest for one tag hash - your suggestion adds unnecessary complexity IMHO

Reply:
Even though `<mirror-dir>/working-dir/operator-catalogs/redhat-operator-index/e3368fb4222d7586a1505748f383fa711b2a2baa231dd02c02b974fd19ade942` is unique (this is the catalog in the imagesetconfig, by digest), `filtered-catalogs/2124f1daa35027087f1c61e6631ae521` represents the hash for a specific filtering, for example:
```
    - catalog: registry.redhat.io/redhat/redhat-operator-index@sha256:e3368fb4222d7586a1505748f383fa711b2a2baa231dd02c02b974fd19ade942 
      packages:
       - name: aws-load-balancer-operator
```

When the user modifies anything in this imagesetconfig stanza, say adding minVersion, this will modify the hash and generate another folder for the filtered catalog  `filtered-catalogs/xyz`.

And since the working-dir is almost never cleaned, this situation can happen.

Reply:
@sherine-k - I have updated to delete the directory for filtered-catalogs on each run - this way we never have t o worry about duplication, and also the change was fairly simple

I have tested using mirror-to-mirror,  mirror-to-disk and disk-to-mirror - 2 runs for each flow using the following imagesetconfig 

```
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v2alpha1
mirror:
  operators:
    - catalog: registry.redhat.io/redhat/redhat-operator-index@sha256:35ea076da9001466a5e4078934fc4a26308488b3424930f8c3a7134f929c5866
       packages:
        - name: rhods-operator
          defaultChannel: fast
          channels:
            - name: fast
              minVersion: 2.17.0
              maxVersion: 2.17.0
```

Then executing with an updated confiig i.e 

```
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v2alpha1
mirror:
  operators:
    - catalog: registry.redhat.io/redhat/redhat-operator-index@sha256:35ea076da9001466a5e4078934fc4a26308488b3424930f8c3a7134f929c5866
      packages:
        - name: rhods-operator
          defaultChannel: fast
          channels:
            - name: fast
              minVersion: 2.17.0
              maxVersion: 2.17.0
        - name: cincinnati-operator
          channels:
            - name: v1

```
Finally I tested all flow by jus tusing a simple tagged catlog i.e

```
kind: ImageSetConfiguration
apiVersion: mirror.openshift.io/v2alpha1
mirror:
  operators:
    - catalog: registry.redhat.io/redhat/redhat-operator-index:v4.18
      packages:
        - name: rhods-operator
          defaultChannel: fast
          channels:
            - name: fast
              minVersion: 2.17.0
              maxVersion: 2.17.0

```

All seems to be working

Reply:
The reason why the filtered-catalog were kept in the directory (working-dir) was to avoid filtering and rebuilding catalogs that were already filtered and built using the same image set configuration. This avoid filtering and rebuilding catalogs every time (it would have impact in enclaves with several nodes).

Cleaning always the folder will make oc-mirror to lose this behavior (which is a good behavior from my perspective since it improves performance and saves time).

An alternative solution for this problem would be selecting the correct digest of the filtered-catalog and use this one instead of the one in the first index [0].